add Slices array type for eachslice/eachrow/eachcol

simonbyrne opened the pull request:
`eachslice`, `eachrow`, `eachcol` (introduced in JuliaLang/julia#29749) now return an `SliceArray` object (along with `Rows`/`Columns` aliases). The main benefit is that it will allow dispatch on the iterator to provide more efficient methods, e.g.
```
sum(A::Rows) = vec(sum(parent(A), dims=1))
```

This will encourage the use of `eachcol`/`eachrow` to resolve ambiguities in user-facing APIs, in particular, the "observations as rows vs columns" problem in the statistics/ML packages.

This also makes `eachslice` work over multiple dimensions.

TODOs:
- [x] Figure out how to handle `eltype`: I'd appreciate some help on how best to do this.

cc: @nalimilan @mschauer @ararslan @kleinschmidt @mbauman @andyferris @bramtayl @arnavs

bramtayl commented:
The way I handle `eltype` in `JuliennedArrays` is with `typeof(@inbounds view(whole, first_slice_dimensions)...)`. `@inbounds` gives the advantage of allowing it to work with empty arrays. If the `view` function is unstable on `whole`, it should just error when a slice doesn't match the type of the first slice. I think the long term solution is to allow for `eltype`-less `AbstractArrays` by making `AbstractArray` just a collection of traits (2.0?).

bramtayl commented:
Also, instead of `sum(A::EachRow) = vec(sum(parent(A), dims=1))`, it probably makes more sense to dispatch to mapreducedims for any reducing `dims` function, so this would be helpful:

```
struct Reduce{Operator}
    operator::Operator
end

sum = Reduce(+)
```

simonbyrne commented:
> Also, instead of `sum(A::EachRow) = vec(sum(parent(A), dims=1))`, it probably makes more sense to dispatch to mapreducedims for any reducing `dims` function

Yes, this was just meant to be an illustrative example.

ararslan commented on base/abstractarraymath.jl:
```diff
@@ -407,6 +407,26 @@ _reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
+
+function iterate(s::EachSlice, state...)
+    r = iterate(s.cartiter, state...)
+    isnothing(r) && return r
```
We should probably use `=== nothing` everywhere instead of `isnothing` because of #27681. (Similarly applies to `nothing`/`isnothing`.)
```suggestion
    r === nothing && return r
```

simonbyrne commented:
Figuring out the appropriate `SubArray` type looks complicated. I'll fallback on the `EltypeUnknown` machinery for now (which matches the current `Generator` behaviour), perhaps figure out something better in future.

andyferris commented:
Cool.

I think there are advantages in preserving the `AbstractArray` nature of the container - e.g. splitting a matrix into view which is a vector of vectors. (Doing the reverse is also super useful sometimes).

bramtayl commented:
Well, just added the capability of processing `dim::Int...` to JuliennedArrays. Should be type stable under constant propagation.

simonbyrne commented:
> Honestly, though, maybe it makes sense just to copy-paste the code from JuliennedArrays and define
> 
> `eachcol(A) = Slices(A, True(), False())`
> `eachrow(A) = Slices(A, False(), True())`
> 
> ? I've seen an uptick in attention to it.

I did consider it, but there were a few things that made me a bit uneasy: e.g. the use of `@inbounds` does open up potential problems if the type can't actually be inferred without creating an instance. It was also quite a bit more code for what I wanted to achieve: as I said in the PR immediate aim was simply to be able to dispatch on the return type of `eachrow`/`eachcol` and be able access the underlying array easily

> I think there are advantages in preserving the `AbstractArray` nature of the container - e.g. splitting a matrix into view which is a vector of vectors. (Doing the reverse is also super useful sometimes).

I agree: the main challenge in doing that is figuring out the `SubArray` element type. That said, I think that can also be done later (it would technically involve changing the type hierarchy, but not in a terribly breaking way).

simonbyrne commented:
One thing that is perhaps worth noting about my PR is that the order of `dims` is important:
```
julia> M = reshape([(1:8)...], 2, 2, 2)
2×2×2 Array{Int64,3}:
[:, :, 1] =
 1  3
 2  4

[:, :, 2] =
 5  7
 6  8

julia> collect(eachslice(M,dims=(2,3)) )
2×2 Array{Any,2}:
 [1, 2]  [5, 6]
 [3, 4]  [7, 8]

julia> collect(eachslice(M,dims=(3,2)) )
2×2 Array{Any,2}:
 [1, 2]  [3, 4]
 [5, 6]  [7, 8]
```
happy to hear people's thoughts on this.

bramtayl commented:
> the use of @inbounds does open up potential problems if the type can't actually be inferred without creating an instance

Can you give more detail? I'll need to figure out a way to fix JuliennedArrays if so.

simonbyrne commented:
> Can you give more detail? I'll need to figure out a way to fix JuliennedArrays if so.

Well, you're using `@inbounds` when the object itself might not be "in bounds", so there is the potential for undefined behaviour. I know you're not actually accessing elements of the view, so it is probably okay, but I'm personally not 100% sure it's safe, so don't want to be the person adding that code to Julia.

nickrobinson251 commented:
bump? :)

simonbyrne commented:
@mbauman any thoughts?

bramtayl commented:
Ref #32940

mbauman commented:
This looks great.  My only hesitation is that we worked quite a bit on performance in https://github.com/JuliaLang/julia/pull/29749 — this looks like it should be good, but have you done any spot checks?

mcabbott commented on base/abstractarraymath.jl:
```diff
@@ -449,9 +480,10 @@ See also [`eachrow`](@ref), [`eachcol`](@ref), and [`selectdim`](@ref).
      This function requires at least Julia 1.1.
 """
 @inline function eachslice(A::AbstractArray; dims)
```
Would it be possible to have an intermediate step to allow overloading of dispatch on `dims`? Motivated by `dims=:μ` type things in NamedDims.
```suggestion
eachslice(A::AbstractArray; dims) = _eachslice(A::AbstractArray, dims)
@inline function _eachslice(A::AbstractArray, dims)
```

MasonProtter commented:
Is there any interestin reviving this?

simonbyrne commented on base/abstractarraymath.jl:
```diff
@@ -487,11 +519,21 @@ julia> collect(eachslice(M, dims=2))
+    iter = CartesianIndices(map(dim -> axes(A,dim), dims))
+    lookup = ntuple(dim -> findfirst(isequal(dim), dims), N)
+
+    # determine element type
+    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : eltype(a), axes(A), lookup)...}
+    hasstride = IndexStyle(A) == IndexLinear() # has linear indexing if parent does
+    E = SubArray{eltype(A),N-M,typeof(A),I,hasstride}
```
@mbauman does this logic make sense? Should it be in a separate function?

mbauman commented on base/abstractarraymath.jl:
```diff
@@ -416,8 +444,13 @@ julia> collect(eachrow(a))
+function eachrow(A::AbstractVecOrMat)
+    iter = CartesianIndices((axes(A,1),))
+    lookup = (1,nothing)
+    EachSlice(A,iter,lookup)
```
Surely this needs a constructor to figure out `E` and `M`?

mbauman commented on base/abstractarraymath.jl:
```diff
@@ -487,11 +525,21 @@ julia> collect(eachslice(M, dims=2))
+    lookup = ntuple(dim -> findfirst(isequal(dim), dims), N)
+
+    # determine element type
+    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : eltype(a), axes(A), lookup)...}
```
We need to explicitly communicate to SubArray that these are going to be slices over the entire dimension (that helps make things linear more frequently).

```suggestion
    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : Base.Slice{eltype(a)}, axes(A), lookup)...}
```

mbauman commented on base/abstractarraymath.jl:
```diff
@@ -487,11 +525,21 @@ julia> collect(eachslice(M, dims=2))
+
+    # determine element type
+    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : eltype(a), axes(A), lookup)...}
+    hasstride = IndexStyle(A) === IndexLinear() # has linear indexing if parent does
```
Continuing the discussion from (the now outdated) https://github.com/JuliaLang/julia/pull/32310#discussion_r540350659, the resulting subarray will only be linear if the slices are all across contiguous dimensions.  Here's a counter-example test-case:

```Julia
julia> A = rand(5,4,3,2); dims=(2,4); N=4; M=2;

julia> typeof(view(A, :, 1, :, 1))
SubArray{Float64,2,Array{Float64,4},Tuple{Base.Slice{Base.OneTo{Int64}},Int64,Base.Slice{Base.OneTo{Int64}},Int64},false}
```

This view is strided but not linear.

simonbyrne replied:
That's unfortunate.

How bad would it be to use `Core.Compiler.return_type` here?

andyferris replied:
It would certainly simplify things - assuming it doesn't make the compiler crash or compilation invalid?

mbauman reviewed (commented):
few comments on the subarray type prediction

mbauman commented on base/abstractarraymath.jl:
```diff
@@ -386,6 +386,34 @@ end
+    r = iterate(s.cartiter, state...)
+    r === nothing && return r
+    (c,nextstate) = r
+    view(s.arr, map(l -> l === nothing ? (:) : c[l], s.lookup)...), nextstate
```
Ah, and `view` isn't guaranteed to return a `SubArray` — we should probably call into the constructor directly.

simonbyrne replied:
I was wondering about this (I know it's not true for e.g. `CuArray`s). Would it make sense to go the other way, i.e. have a function that returns the type of whatever the `view` would be?

mbauman replied:
Yeah, we could do that.  I think I'd lean towards doing it in the value domain, though, as then we can fall back to a simple `typeof(view(...))`.  The whole point of views is that they should be cheap (or even no-cost) to construct (especially if not used).

andyferris replied:
The problem here is in the empty case - we still need to return something.

bramtayl commented on base/abstractarraymath.jl:
```diff
@@ -487,11 +525,21 @@ julia> collect(eachslice(M, dims=2))
+
+@inline _eachslice(A,dim::Integer) = _eachslice(A,(dim,))
+@inline function _eachslice(A::AbstractArray{N}, dims::NTuple{M,Integer}) where {N,M}
+    for (i,dim) in enumerate(dims)
```
Worth noting that you could do this with recursion instead to get type stability under constant propagation, c.f. the current version of JuliennedArrays.

simonbyrne replied:
Good point.

andyferris commented on base/abstractarraymath.jl:
```diff
@@ -386,6 +386,34 @@ end
 
 end#module
 
+struct EachSlice{E,M,A,I,L} <: AbstractArray{E,M}
```
I love that this is an `AbstractArray`. :)

I note that the majority of our concrete array types tend to be `XxxArray`, so as a suggestion we could name this as `SlicedArray`?

chriselrod replied:
What's the advantage of being distinct from `SubArray` with `Base.Slice` indices?

mbauman replied:
This means something different?  SubArrays iterate over their elements; this iterates over whole slices (SubArrays themselves).  I don't think you can bend a SubArray by itself into this behavior.

chriselrod replied:
Sorry for the noise, I had a misunderstanding.

nalimilan commented:
Is anything blocking progress on this PR? It would be really useful for several applications and we keep mentioning it. :-)

simonbyrne commented:
My time?

I think it should be a subtype of `AbstractArray`, but I don't have the bandwidth to work on this at the moment. If anyone else is interested, I would heartily encourage them to take it on.

andyferris commented:
I had been under the impression that this was always stuck on a decision with the eltype calculation? For example, we can't assume `view` even returns a `SubArray`.

I can see a few approaches:

1. Always use `return_type(view, ...)`.
2. Use `typeof(view(...))` unless it is empty, in which case use `return_type(view, ...)`
3. Use `typeof(view(...))` unless it is empty, in which case use eltype `Union{}` or something.

Restricting `return_type` usage to the empty case (2) doesn't appear to help anything over (1), as far as I can tell. (We do this in `collect`, nonetheless).

In (3) the output would always be a `Union{EachSlice{...}, EachSlice{Union{}}`. Having `Union{}` here has the advantage that after using iteration or getindex on this downstream, at least the types should re-converge correctly (we will benefit from `Union{A, Union{}} == A`), and possibly the compiler is smart enough not to recheck the types on every iteration.

My personal preference is (1) - until/unless we figure out how to entirely remove `return_type` from `collect` in the empty case, at which point we can re-audit `Base`.

andyferris commented:
> If anyone else is interested, I would heartily encourage them to take it on.

Oh I probably should have mentioned above, if we can resolve that I may be able to help out here.

simonbyrne commented:
> My personal preference is (1) - until/unless we figure out how to entirely remove `return_type` from `collect` in the empty case, at which point we can re-audit `Base`.

I agree as well.

bramtayl commented:
I think the first best solution would be something like this:

1) Make a lazy AbstractArray where each element is a set of sliced indices, so for example:

```
sliced_indices(A, along_rows)[1] = (1, 1:n)
```

2) Make `Iterators.map` return an `eltype`-less AbstractArray

3) Make a wrapper function for `Iterators.map`ing `view` over sliced indices.

I'd be happy to help out if we go this route. I don't see any of the other options as better than what already exists in JuliennedArrays?

andyferris commented:
What is an “eltypeless“ `AbstractArray{T, N}`?

andyferris commented:
FYI I think eventually we might want to remove those abstract type parameters from `AbstractArray` and replace them with traits, so we can have (especial lazy) arrays that are ElTypeUnknown, but it’s a breaking change and a lot of work and well beyond the scope here.

andyferris commented:
> I don't see any of the other options as better than what already exists in JuliennedArrays?

@bramtayl I see the usage of `typeof(@inbounds view(...))` in JuliennedArrays. That _is_ a clever way of dealling with the empty case, but unfortunately it's not going to work with `--check-bounds=yes` and otherwise it won't _necessarily_ work with every custom `view` container other people create that don't have to listen to `@inbounds`.

```
ferris@andy-laptop:~/dev/ELARACore (edk)$ julia --check-bounds=yes
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using JuliennedArrays

julia> Slices([1 2; 3 4], 1)
2-element Slices{SubArray{Int64,1,Array{Int64,2},Tuple{Base.OneTo{Int64},Int64},true},1,Array{Int64,2},Tuple{True,False}}:
 [1, 3]
 [2, 4]

julia> Slices(zeros(0, 2), 1)
2-element Slices{SubArray{Float64,1,Array{Float64,2},Tuple{Base.OneTo{Int64},Int64},true},1,Array{Float64,2},Tuple{True,False}}:
 0-element view(::Array{Float64,2}, Base.OneTo(0), 1) with eltype Float64
 0-element view(::Array{Float64,2}, Base.OneTo(0), 2) with eltype Float64

julia> Slices(zeros(2, 0), 1)
ERROR: BoundsError: attempt to access 2×0 Array{Float64,2} at index [1:2, 1]
Stacktrace:
 [1] throw_boundserror(::Array{Float64,2}, ::Tuple{Base.OneTo{Int64},Int64}) at ./abstractarray.jl:541
 [2] checkbounds at ./abstractarray.jl:506 [inlined]
 [3] view at ./subarray.jl:158 [inlined]
 [4] Slices at /home/ferris/.julia/packages/JuliennedArrays/furG7/src/JuliennedArrays.jl:131 [inlined]
 [5] Slices(::Array{Float64,2}, ::Int64) at /home/ferris/.julia/packages/JuliennedArrays/furG7/src/JuliennedArrays.jl:164
 [6] top-level scope at REPL[4]:1
```

I would rather avoid adding any "undefined" behavior to `Base`. The behaviour under `--check-bounds=yes` has to be considered the correct one, and generally any `@inbounds` annotations that could violate that are left to end-users to add (we use them internally only to avoid a _second_ check of the same thing, and use `@propagate_inbounds` to let the user "reach in" when they want).

One could error out always on the empty case but I also find that unsatisfying.

bramtayl commented:
Right, I'm not saying the eltype calculation in JuliennedArrays is great, and happy to take a PR if someone has a better idea, I'm just saying for a greedy language like Julia its a good idea to keep the best solution in mind (even if it's far off)

andyferris commented:
Yes, I see - we all look forward to the glorious eltypeless abstract array future, then! 🤣

In the shorter term I might rebase and modify this and see how `return_type` behaves in practice, when I get some free cycles.

simonbyrne commented:
I think I figured out a basic implementation here:
https://github.com/simonbyrne/SlicedArrays.jl
It's able to handles some basic test cases. Any ideas for good benchmarks?

andyferris commented:
Awesome - and I thought you were too busy :wink:

> Any ideas for good benchmarks?

I think just the basic "split apply combine" pattern is what we are looking for. Like creating a vector containing the mean of each row of a matrix - `r = rand(n, m); @btime mean.(eachrow($r))`. It should have similar performance to a hand-written loop (using `@inbounds`), and hopefully much better than materializing the rows like `[mean(r[i, :]) for i in axes(r)[1]]`

simonbyrne commented:
One challenge I'm having is getting constant propagation to work across keyword arguments.
```
_eachrow(X) = SlicedArrays._eachslice(X,(1,))
```
will infer the correct return type, but
```
_keachrow(X) = SlicedArrays.eachslice(X,dims=(1,))
```
will not, despite being an inlined call to the former.

(the current implementation doesn't do this either, but it does seem like a desirable feature)

andyferris commented:
Yes I’ve had trouble like that before, too. It’s a shame as often keyword arguments make for a better API.

simonbyrne commented:
The other question is should it support optional keeping of inner or outer dimensions (so we could implement e.g. `mapslices` functionality).

simonbyrne committed:
Add SliceArray type for eachslice/eachcol/eachrow

simonbyrne commented:
Okay, I think this is ready for feedback:
- I still can't figure out the constant propagation with `eachslice`, but it `eachrow`/`eachcol` are type-inferrable.
- I went with `SliceArray` rather than `SlicedArray`: my rationale is if it is an "array of slices", then `eachcol(::Matrix)` gives you a "vector of slices". If it was a "sliced array", then `eachcol(::Matrix)` gives you a "sliced matrix". Happy to change it though.

andyferris commented:
Ah, bikeshedding, the important work :wink: It's so hard to avoid.

Typically the prefix is an adjective describing the array itself (or what happened to it for lazy views, past tense), not so much the elements *per se*.

* `SparseArray`
* `ReshapedArray`
* `PermutedDimsArray`
* `StaticArray`
* `SubArray`

There is one I see that uses a verb form.

* `ReinterpretArray`

But it's still not "an array of reinterprets".

`BitArray` could go either way - describing the layout as being made up of bits, like the way `SparseArray` describes it's layout as sparse. You _might_ say it's elements are bits, like you'd say the `SliceArray` elements are slices.

Personally - our own style guide clearly using suggests generic naming generic functions as verbs, so that would be `slice(...)` (rather than `eachslice`) that is lazy - and therefore produces a `SlicedArray`?

But it's not the most important question in the scheme of things :)

simonbyrne commented:
If we go with `SlicedArray`, what would we call `RowSliceVector`? (since giving that a name was the original point of this PR): would it be a `RowSlicedVector` or a `RowSlicedMatrix`?

andyferris commented:
Is `RowVector` ok?

andyferris commented:
I suppose that’s like `SliceArray`.

andyferris commented:
Naming things is hard...

andyferris commented:
`RowVector`, `ColumnVector` and `SliceArray` do seem simple, clear and self consistent.

nalimilan commented:
`RowVector` and `ColumnVector` would be confusing IMHO: these terms are usually used to refer to single-row and single-column vectors (see the old https://github.com/JuliaLang/julia/issues/23018). `RowSliceVector` and `ColumnSliceVector` sound more explicit.

mcabbott commented:
Isn't "row vector" quite often used for what Julia calls `AdjointAbsVec`? 

What examples are there of when this alias would be used? One I can think of is `Distances.pairwise(::Metric, ::EachRow)` which really wants a matrix to work with, and you are telling it how to interpret that. (If it was really going to work with slices, it would not care where they came from, only about `AbstractVector{<: AbstractVector}`.) Perhaps that points towards e.g. `SlicedArray`, `SlicedMatrixRows`, `SlicedMatrixColumns`, referencing the parent? Those are also tab-completion-friendly.

simonbyrne commented:
How about `RowsVector`/`ColsVector`?

simonbyrne commented:
> One I can think of is `Distances.pairwise(::Metric, ::EachRow)` which really wants a matrix to work with, and you are telling it how to interpret that.

That is exactly the use-case I had in mind. We could go with `EachRow` or `EachColumn`?

simonbyrne commented:
Or even just `Rows` and `Columns`?

mcabbott commented:
Re-using the function names does seem pretty unambiguous, then there's no relation like `view` <-> `SubArray` to memorise.

`Rows` / `Columns` / `Slices` are pretty, but perhaps too short for Base? They would also be nice as function names; it looks like the `each` was added to indicate generators, and I guess we are stuck with it.

simonbyrne committed:
rename aliases to Rows/Columns

andyferris commented:
> there's no relation like view <-> SubArray to memorise.

Another point is for both `SubArray` and `SliceArray` as a user I don't tend to need to know what they are called - just the generic function that creates them. (Maybe when contributing to `Base` or explaining lazy containers to someone it is useful to know).

mcabbott commented:
Indeed, I guess the main time I see `SubArray` is in stack traces.

If dispatching on `::Rows` etc. becomes a common interface alternative to taking a keyword `dims`, then more people may interact with these, like the `Distances.pairwise(::Metric, ::EachRow)` example. Good that the parent is the first parameter, BTW, `::Rows{<:StridedMatrix}` etc.

simonbyrne committed:
fix field name

simonbyrne committed:
fix test

simonbyrne committed:
more test fixes

simonbyrne commented:
Any more bikeshedding on names? Currently we have:
- `SliceArray` as the generic object for `eachslice`
- `Rows` and `Columns` as the aliases for `eachrow`/`eachcol`

This will probably require triage input, both due to addition of extra exports, and use of `Base._return_type`.

simonbyrne commented:
@mbauman @andyferris any more comments?

andyferris reviewed (approved):
This looks good to me! (I'll leave any remaining bikeshedding to triage). Thanks @simonbyrne

StefanKarpinski commented:
Call it `Slices`?

JeffBezanson commented:
:+1: 
Can we use `:` instead of `nothing`? It seems to make sense and would be much more aesthetic :)

simeonschaub commented:
I think it would be good to run PkgEval before merging, because people might be at least indirectly relying on these returning a `Generator`.

mbauman commented:
> Can we use `:` instead of `nothing`?

It's a little weird.  Does that `L` parameter need to be a parameter at all?  It's effectively a transformer that specifies how a given index changes into the view index.  E.g., given 2d index `I`, generate out of it `(:, I[2], :, I[1])`.  That operation is currently encoded as `(nothing, 2, nothing, 1)`.  What is this operation?  Is there a better first-class way of representing it?

mbauman commented on base/slicearray.jl:
```diff
@@ -0,0 +1,224 @@
+size(s::SliceArray, dim) = size(s.cartiter, dim)
+length(s::SliceArray) = length(s.cartiter)
+
+@inline function _slice_index(s::SliceArray{N,L,P,CI,S}, I...) where {N,L,P,CI,S}
+    c = s.cartiter[I...]
```
Is there any case where this _isn't_ a no-op?  Or, rather, can't we just do:

```suggestion
@inline function _slice_index(s::SliceArray{N,L,P,CI,S}, c::Vararg{Int,N}) where {N,L,P,CI,S}
```

We'd of course still need to store the axes (instead of the cartiter), but shouldn't this be exactly the same performance?  And simpler?

simonbyrne replied:
Good point. I'm not sure why I did it this way...

mbauman replied:
I totally get it: you want an iterator over the locations; that's the `CartesianIndices`.  Arrays just bring those along as part of their indexing (if they're not IndexLinear, anyhow).  Took me a minute to see, too.

simonbyrne committed:
use axes instead of cartiter

simonbyrne committed:
rename SliceArray => Slices

simonbyrne committed:
use : as sentinel instead of nothing

simonbyrne commented:
Okay, I've addressed the comments above (other than the PkgEval one).

simonbyrne commented:
> It's a little weird. Does that `L` parameter need to be a parameter at all? It's effectively a transformer that specifies how a given index changes into the view index. E.g., given 2d index `I`, generate out of it `(:, I[2], :, I[1])`. That operation is currently encoded as `(nothing, 2, nothing, 1)`. What is this operation? Is there a better first-class way of representing it?

I agree: I tried to localize its use to a single function (`_slice_index`) so that it should be possible to switch it out for something else.

mbauman commented:
Even as it currently stands, it might make sense to store _the value_ of `L` in the struct as it'll otherwise generate more specializations than necessary.  That is, I think we can deal with `(:, 1, :, 2)` using the same specialization as `(:, 2, :, 1)`, yeah?  Then the parameter would be encoded as `Tuple{Colon, Int, Colon, Int}`.

simonbyrne committed:
typos

simonbyrne committed:
more fixes

mbauman commented on base/slicearray.jl:
```diff
@@ -212,7 +212,7 @@ IteratorSize(::Type{Slices{N,L,P,AX,S}}) where {N,L,P,AX,S} = HasShape{N}()
 size(s::Slices) = map(length, s.axes)
 
-@inline function _slice_index(s::Slices{N,L}, c::Vararg{Int,N}) where {N,L}
+@inline function _slice_index(s::Slices{N,L}, c...) where {N,L}
```
You'll need to restrict the `getindex` below (it's _just_ out of reach of the editor here)

simonbyrne replied:
I'm not sure I understand: what should it be?

mbauman commented on base/slicearray.jl:
```diff
@@ -0,0 +1,222 @@
+    return map(l -> l === (:) ? (:) : c[l], L)
+end
+
+getindex(s::Slices, I...) = view(s.parent, _slice_index(s, I...)...)
```
Here:

```suggestion
getindex(s::Slices{N}, I::Vararg{Int,N}...) where {N} = view(s.parent, _slice_index(s, I...)...)
```

This way Julia will ensure we always get `N` indices as `_slice_index` expects

simonbyrne committed:
fix indexing dispatch

simonbyrne committed:
move to slicemap to a field

simonbyrne commented:
> Even as it currently stands, it might make sense to store _the value_ of `L` in the struct as it'll otherwise generate more specializations than necessary. That is, I think we can deal with `(:, 1, :, 2)` using the same specialization as `(:, 2, :, 1)`, yeah? Then the parameter would be encoded as `Tuple{Colon, Int, Colon, Int}`. I've also just thought about directly embedding a function that does this operation... but the subarray type prediction gets much messier that way.

I was skeptical that would work, but it does seem to work (and give good inference results)! I've updated the PR, and also updated https://github.com/simonbyrne/SlicedArrays.jl (which is the same code) if you want to try it out.

simonbyrne committed:
doc fixes

simonbyrne commented on base/slicearray.jl:
```diff
@@ -0,0 +1,219 @@
+
+[`parent(S::Slices)`](@ref) will return the parent array.
+"""
+struct Slices{N,P,SM,AX,S} <: AbstractArray{S,N}
```
@mbauman One other thing I'd appreciate your opinion on is the parameter order: I pushed the element type `S` to the end, since it is typically complicated, and also not that important. I originally had `N` as the first one since it made some constructors easier, but I don't think that is really required any longer.

mbauman replied:
Yeah, I can see this both ways.  I think I'd lean towards just doing the `Slices{S, N, ...}` thing to fit with convention.  It's gonna be a mouthful of a type in any case, might as well match convention.

simonbyrne replied:
It probably does make sense to have element type after parent type though, as it makes using the `Rows`/`Colums` aliases easier, i.e. `f(X::Rows{Matrix{Float64}}) = ...`

simonbyrne committed:
rearrange parameters

simonbyrne commented:
bump?

simonbyrne commented:
@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/32310#issuecomment-882871376) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/2c89586_vs_d786a90/report.md).

johnnychen94 commented:
> Call it Slices?

Because `Base.Slice` exists with a different meaning (represents the range), I think `SliceArray` (represents the value) is more accurate here.

simonbyrne commented:
So, any final bikeshedding on names?

nalimilan commented on base/slicearray.jl:
```diff
@@ -0,0 +1,222 @@
+        return Slices(A, slicemap, ax)
+    end
+end
+@inline function _eachslice(A::AbstractArray{T,N}, dim::Integer, drop::Bool) where {T,N}
```
```suggestion
@inline function _eachslice(A::AbstractArray, dim::Integer, drop::Bool)
```

nalimilan commented on base/slicearray.jl:
```diff
@@ -0,0 +1,222 @@
+ 4  5  6
+ 7  8  9
+
+julia> S = eachslice(M,dims=1)
```
```suggestion
julia> S = eachslice(M, dims=1)
```

nalimilan commented on base/slicearray.jl:
```diff
@@ -0,0 +1,222 @@
+ 2
+ 3
+
+julia> T = eachslice(M,dims=1,drop=false)
```
```suggestion
julia> T = eachslice(M, dims=1, drop=false)
```

nalimilan commented on base/slicearray.jl:
````diff
@@ -0,0 +1,222 @@
+```
+"""
+eachcol(A::AbstractMatrix) = _eachslice(A, (2,), true)
+eachcol(A::AbstractVector) = eachcol(reshape(A, size(A,1), 1))
````
```suggestion
eachcol(A::AbstractVector) = eachcol(reshape(A, size(A, 1), 1))
```

nalimilan reviewed (commented):
I like the names! Just a few suggestions to improve the consistency of docstrings.

simonbyrne committed:
Apply suggestions from code review

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

simonbyrne commented:
For statistics, it is conceptually a more understandable way than using `dims` keywords arguments to denote whether "observations are rows" or "observations are columns", see https://discourse.julialang.org/t/about-the-convention-for-the-design-matrix-row-as-observation-or-column-as-observation/20080

bkamins commented:
Statistics was exactly the reason I have asked. See https://discourse.julialang.org/t/julia-stats-data-ml-expanding-usability/67948. AFAICT it is currently not settled what is the best design of statistics, so I thought that it is better not to bake-in into Julia Base some decisions that would constrain the general design later.

In particular a general thing that people often want is to support Tables.jl tables in statistical ecosystem, and Tables.jl defines `Tables.Columns` also.

simonbyrne commented:
Yeah, the naming is something i am concerned about, which why i haven't merged it. @quinnj any thoughts?

bkamins commented:
Maybe names `ArrayRows`, `ArrayColumns` (and potentially also `ArraySlices`) would be more explicit?

devmotion commented on base/slicearray.jl:
```diff
@@ -0,0 +1,223 @@
+    return map(l -> l === (:) ? (:) : c[l], s.slicemap)
+end
+
+getindex(s::Slices{P,SM,AX,S,N}, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
+    view(s.parent, _slice_index(s, I...)...)
+setindex!(s::Slices{P,SM,AX,S,N}, val, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
+    s.parent[_slice_index(s, I...)...] = val
```
I think it would be useful if the `@inbounds` context would be forwarded here:
```suggestion
Base.@propagate_inbounds getindex(s::Slices{P,SM,AX,S,N}, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
    view(s.parent, _slice_index(s, I...)...)
Base.@propagate_inbounds setindex!(s::Slices{P,SM,AX,S,N}, val, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
    s.parent[_slice_index(s, I...)...] = val
```

simonbyrne committed:
Update base/slicearray.jl

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

oschulz commented:
Could we add an `AbstractArraySlice` instead of `Slices` being a subtype of `AbstractArray` directly, with the contract that `parent` will return the unsliced array?

I'd like to make [`ArraysOfArrarys.ArraysOfSimilarArrays`](https://github.com/JuliaArrays/ArraysOfArrays.jl) a subtype of it. `ArraysOfSimilarArrays` is similar to `Slice` but can't be fully replaced by it, as functionality is a bit different: It's limited to being sliced between leading and trailing dimensions, but offers `append!` and similar functions in return.

With a `AbstractArraySlice` supertype, packages like Distributions could take advantage of the fact that a flat array is available (e.g. when calculating a broadcasted PDF or likelihood) in a way that leaves room for other sliced-array implementations without dependency problems (see JuliaArrays/ArraysOfArrays.jl#20).

oschulz commented:
> With an AbstractSlices supertype

In addition to that, might there also be room here for supporting `vcat` for `Slices` that are semantically vectors-of-arrays?

mcabbott commented on base/slicearray.jl:
```diff
@@ -0,0 +1,223 @@
+ 2
+ 3
+
+julia> T = eachslice(M, dims=1, drop=false)
+3×1 Slices{Matrix{Int64}, Tuple{Int64, Colon}, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, 2}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
```
To explain what `dims` does, perhaps this should show a higher-dimensional example. They are messy to print but showing just the sizes ought to make it clear, something like this:
```suggestion
julia> sizes(x) = (element = size(first(x)), container = size(x));

julia> eachslice(M, dims=1, drop=false) |> sizes
(element = (4,), container = (3, 1))

julia> A = rand(5, 7, 11, 13);

julia> eachslice(A; dims=(2,4)) |> sizes
(element = (5, 11), container = (7, 13))

julia> eachslice(A; dims=(4,2)) |> sizes
(element = (5, 11), container = (13, 7))

julia> eachslice(A; dims=(4,2), drop=false) |> sizes
(element = (5, 11), container = (1, 7, 1, 13))
```
(where I've changed M to be 3×4 so that size 3 is unambiguous, maybe 2×3 would be better.)

nalimilan commented:
So this missed 1.8. Can we aim at merging the PR in time for Julia 1.9? :-)

According to votes above, it seems like the names `RowSlices`/`ColumnSlices` are the most popular. Adding an `AbstractSlice` supertype can happen after merging the PR, so let's not let more discussions get in the way?

oschulz commented:
Could we have an AbstractSlices supertype as part of this, for packages to extend (see my comment above)?

devmotion commented:
Bump 🙂

It would be great to get this into 1.9. I agree with @nalimilan, to avoid further delay the super type could be discussed and added in follow-up PRs.

oschulz commented:
>  I agree with @nalimilan, to avoid further delay the super type could be discussed and added in follow-up PRs.

I would argue that adding the super-type doesn't need to delay this. And if this becomes part of Julia done without such a super-type that would make life hard on packages like ArraysOfArrays - since there would be an official nested array mechanism that's difficult to be compatible with and extend for arrays-of-arrays implementations that need to offer more extensive features. And once this is in and in use, it may be harder to convince people to touch it again and add a super-type - which would then require applications to depend on an even higher Julia version.

nalimilan committed:
Rename types, add supertype, some minor fixes

nalimilan commented:
OK, I've rebased against latest master, renamed types to `RowSlices`/`ColumnSlices` and added an `AbstractSlices` supertype. I also made a few minor changes which I hope everybody is OK with.

@oschulz Does this suit your needs?
@simonbyrne Are you OK with these changes?
@mbauman Could you confirm that the PR looks good now?

devmotion commented on base/slicearray.jl:
```diff
@@ -0,0 +1,234 @@
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
```
I guess this should refer to 1.9 as in the other docstrings?

```suggestion
!!! compat "Julia 1.9"
     Prior to Julia 1.9, this returned an iterator.
```

nalimilan committed:
Fix doctest

nalimilan committed:
Fix version in docstring

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

oschulz commented:
> @oschulz Does this suit your needs

Thanks a lot for adding `AbstractSlices`!

One question: Is there a standard function to get the sliced dimensions, like as a `Val((3,1,5))` (or similar), given an `AbstractSlices` instance? I think that would be necessary for some use cases. When getting the flat array via `parent`, it will often be necessary to get the "dimension order", so to speak.

mcabbott commented:
Re the supertype, the [original comment](https://github.com/JuliaLang/julia/pull/32310#issuecomment-989754172) describes something which is still backed by a solid array. Are all expected to be like that, `parent` should always make sense? 

If so, then it would probably make sense to encode the "slicemap" in the abstract type, and provide a function to get it. Also, perhaps `const ColumnSlices` should then refer to the abstract type, since anyone dispatching on it would probably accept any such type.

(This really should not delay it, but I suggested [here](https://github.com/JuliaLang/julia/pull/32310#discussion_r772488699) that the examples could more clearly show what something like `dims=(4,2), drop=false` is going to produce.)

nalimilan commented:
I knew this `AbstractSlices` supertype was going to slow down the PR. I almost regret adding it! :-) Can we defer this to a later PR? Feel free to file an issue to start discussing it now though.

Do you have any objection to the current state of the PR that would have to be fixed before merging?

simonbyrne committed:
Merge branch 'master' into sb/eachslice-type

simonbyrne committed:
Update base/abstractarraymath.jl

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

simonbyrne commented:
Thanks @nalimilan: I'm happy with the changes!

oschulz commented:
> Just a getslicemap function, pretty please? It would be trivial for the existing Slices type, right?

I think just

```julia
"""
    AbstractSlices{S,N} <: AbstractArray{S,N}
Supertype for arrays of slices into a parent array over some dimension(s),
returning views that select all the data from the other dimensions.
`parent` will return the parent array. `getslicemap` will return the
map of sliced dimensions
"""
abstract type AbstractSlices{T,N} <: AbstractArray{T,N} end

"""
    getslicemap(A::AbstractSlices)

Return the map of sliced dimensions in respect to `parent(A)`.
"""
function getslicemap end

# ...

getslicemap(A::Slices) = A.slicemap
```

would already be sufficient.

oschulz commented:
If the risk of adding `getslicemap` now seem to high, we could add it in the next round and "back-port" it via Compat, right? 

It would actually be really nice to add the `AbstractSlices` type to Compat as well. Would be weird if (e.g.) `AbstractArrays.ArrayOfSimilarArrays` would have different supertypes depending on Julia version.

oschulz commented:
> If the risk of adding getslicemap now seem to high, we could add it in the next round and "back-port" it via Compat, right? It would actually be really nice to add the AbstractSlices type to Compat 

If that is acceptable, I would say let's merge this PR now. I'll create a super-lightweight package `AbstractArraysOfArrays` to explore additional interface like `getslicemap` and functionality to reconstruct a slices from a flattened array, so we can evaluate these in practice in actual packages. Then can move them (or part of them) to the language and Compat in another round.

Ok?

oschulz commented:
Assuming Compat will backport `AbstractSlices` and define `parent` for `Compat.EachRow` and `Compat.EachCol` in JuliaLang/Compat.jl#663:

[AbstractArraysOfArrays.jl](https://github.com/JuliaArrays/AbstractArraysOfArrays.jl) [WIP, alternate name suggestions welcome] would define


```julia
const SlicedArrayLike{...} = Union{Compat.EachRow{...},Compat.EachCol{...},Compat.AbstractSlices{...}}

slicemap(A::Compat.EachRow) = ...
slicemap(A::Compat.EachCol) = ...

@static if VERSION >= ...
    slicemap(A::Slices) = A.slicemap
end
```

I'll make sure AbstractArraysOfArrays will have almost zero load time. Packages (I think there a lot of potential in StatsBase, Distributions, Bijectors, possibly Flux and quite a few other) can then begin adding accelerated methods for sliced arrays (should also help making some areas more GPU-compatible). By using `AbstractArraysOfArrays.SlicedArrayLike` "client" packages can to this without depending on non-LTS Julia versions or scattering `if VERSION` through their code. As things mature, we can make a second round of this topic in the language itself and move stuff over from `AbstractArraysOfArrays` to Base and Compat.

Having `AbstractSlices` in Base and Compat will be key to this, though, since the typedef `SlicedArrayLike` can't be extended directly by packages. But JuliennedArrays.jl, ArraysOfArrays.jl and possibly others can subtype `(Base|Compat).AbstractSlices`, so everything will be a `SlicedArrayLike` - including the output of Julia LTS `eachcol` and `eachrow`.

Does this sound like a good plan to move forward?

oschulz commented:
Here's an example statistics use case for the proposal above (merge PR as-is, add `AbstractSlices` to JuliaLang/Compat.jl#663 and establish a package like AbstractArrayOfArray to explore additional interfaces/functionality for sliced arrays like `getslicemap`):

Given

```julia
using LinearAlgebra, Distributions
Σs = rand(5,5)
d = MvNormal(randn(5), Σs * Σs')
xs_c = eachcol(rand(d, 100))
xs_r = eachrow(rand(d, 100)')
```

we currently need to `collect`, because `loglikelihood` currently only supports `AbstractArray{<:AbstractArray{<:Real, N}}`:

```julia
loglikelihood(d, collect(xs_c))
loglikelihood(d, collect(xs_r))
```

Now we'd be able to define

```
using AbstractArraysOfArrays: SlicedArrayLike, getslicemap

function Distributions.loglikelihood(
    d::Distribution{ArrayLikeVariate{N}}, xs::SlicedArrayLike,
) where {N}
    ll_impl(d, parent(xs), getslicemap(xs))
end

function ll_impl(d::Distribution{ArrayLikeVariate{N}}, X::AbstractArray{<:Real,M}, ::Tuple{Colon,Int64}) where {N,M}
    # Currently not super-fast because the generic `loglikelihood` implementation
    # operates col-by-col, but could be specialized for MvNormal to use `d.Σ.chol.L \ X`
    # and the like:
    loglikelihood(d, X)
end

function ll_impl(d::Distribution{ArrayLikeVariate{N}}, X::AbstractArray{<:Real,M}, ::Tuple{Int64,Colon}) where {N,M}
    # Could be specialized for MvNormal, see above:
    loglikelihood(d, X')
end
```

This could would work without `collect` now, and could become very efficient (and even GPU-compatible), from
Julia v1.6 upwards:

```julia
loglikelihood(d, xs_c)
loglikelihood(d, xs_r)
```

nalimilan committed:
Update NEWS.md

nalimilan commented:
Failures also appear in other PRs so they seem unrelated. Merging.

nalimilan commented:
@oschulz Maybe rather than creating an interface package just for this, we could say (in Base) that the `AbstractSlices` interface requires `x.slicemap` to be defined? Custom types can either have this field, or implement it using `getproperty`.

oschulz commented:
> @oschulz Maybe rather than creating an interface package just for this, we could say (in Base) that the AbstractSlices interface requires x.slicemap to be defined? Custom types can either have this field, or implement it using getproperty.

Isn't that a bit "hacky"? It would also be a cumbersome for sliced-array types to implement and would make it impossible to take advantage of dispatch. Also, refitting this to the legacy `eachrow`, `eachcol` and `eachslice` seems ugly.

I think some additional (not too much) interfaces will be required in `AbstractSlices` as well - in particular something that goes beyond `parent`, like a `flat_A, re_f = flatten_with_reconstruct(A)` which will return both the flattened array and a function to reconstruct a similar one. This will be important to have so code can efficiently implement broadcasted operations (e.g. transforming a vector-of-sample-vectors to a new vector-of-sample-vectors, both backed by flat arrays under the hood) without depending on various nested-array-implementations.

oschulz commented:
> I think some additional (not too much) interfaces will be required in an abstract arrays-of-arrays package 

Proposal: I'll flesh out such a super-lightweight package (working title AbstractArraysOfArrays.jl, should be ready soon), prepare a few demo use-cases, and invite proposals/contributions for the final package name and interface before registering?

nalimilan commented:
Maybe this could live in ArrayInterfaces.jl then. That package used to be quite heavy but it's currently reworked to be more lightweight (https://github.com/JuliaArrays/ArrayInterface.jl/issues/211). My point is to avoid scattering interfaces over many small packages.

oschulz commented:
> Maybe this could live in ArrayInterfaces.jl then.

Sure, it it becomes lightweight enough (for packages like StatsBase - `var` and `cov` could certainly profit - to take it on as a dependency). I was planning to rename AbstractArraysOfArrays.jl to SlicedArrayInterface.jl anyway, to stress that it's not specific to ArraysOfArrays.jl - adding it to a lightweight ArrayInterfacesCore.jl instead would make sense.

@ChrisRackauckas, @Tokazama, would there be space for a lightweight interface for sliced arrays in ArrayInterfacesCore.jl,
like `getslicemap` and `flatten_and_reconstruct`? I would be very tiny, load-time wise.

> My point is to avoid scattering interfaces over many small packages.

I guess we should strive for a balance - not too many, but also not too few and too heavy ones. With InverseFunctions.jl and ChangesOfVariables.jl for example, me and @devmotion decided to go for two separate interface packages in the end - looking back I think thats was the right decision. But in this case, it also agree would fit very well into the scope of ArrayInterfacesCore.jl.

Proposal: I implement the interface in an unregistered SlicedArrayInterface.jl (should have this done in a few days), with a demo use case for StatsBase, Distributions and Bijectors. I think I can make a convincing demo for each of them. I'll invite everyone here and the ArrayInterface maintainers (@ChrisRackauckas, @Tokazama ?) for comments on that sliced array interface. And we'll ask the maintainers of packages like the ones mentioned if ArrayInterfaceCore.jl would be an acceptable dependency.

Tokazama commented:
> Proposal: I implement the interface in an unregistered SlicedArrayInterface.jl (should have this done in a few days), with a demo use case for StatsBase, Distributions and Bijectors. I think I can make a convincing demo for each of them. I'll invite everyone here and the ArrayInterface maintainers (@ChrisRackauckas, @Tokazama ?) for comments on that sliced array interface. And we'll ask the maintainers of packages like the ones mentioned if ArrayInterfaceCore.jl would be an acceptable dependency.

Once you have the concept in place I would gladly review it. I can see it being beneficial to have this be in a more general interface package, but I can't make any promises as to whether it will be integrated directly into the ArrayInterface repo until everyone has a chance to review the actual code.

oschulz commented:
> Once you have the concept in place I would gladly review it [...] until everyone has a chance to review the actual code.

Thanks - I definitely want to include everyone interested in the design/discussion, so we end up with something with broad support in the Julia community.

oschulz commented:
@torfjelde and @ToucheSir, I hope such a sliced array interface would also be very useful regarding https://github.com/TuringLang/Bijectors.jl/discussions/178 .

RoyiAvital commented:
Maybe add to documentation an example to use the new multiple dimensions option which was added?

oschulz commented:
> Once you have the concept in place I would gladly review it.

ArrayInterfaceCore is released, so instead of prototyping the sliced array interface standalong, I'll prototype it as an ArrayInterfaceCore PR - Ok @Tokazama?

Tokazama commented:
That seems fine. We still need to set some standards for what goes in core vs other modules because right now there are some things that are left out because of load time issues and not because they don't fit conceptually there there.

oschulz commented:
Thanks @Tokazama I'll give it a go and try to keep it extremely lightweight.

Tokazama commented:
If we have to reconsider which part of the repo it goes in or rework it a bit that shouldn't be too big of a deal. There are a lot of moving pieces right now and I hate for that complexity to discourage meaningful contributions.

oschulz commented:
@Tokazama I'll try to have a PR draft ready within the next days.

oschulz commented:
The current implementation of `Slices` doesn't check indices on single-element dimensions generated by `drop = false`:

```julia
julia> A = eachslice(rand(2,3), dims = 2, drop = false); axes(A)
(Base.OneTo(1), Base.OneTo(3))

julia> A[2,3]
2-element view(::Matrix{Float64}, :, 3) with eltype Float64:
 0.5964895828663271
 0.4755380804162148
```

Should we add `@boundscheck checkbounds(...)` to `getindex(::Slices, ...)`?

mcabbott commented:
Relatedly, should the new axis be `unitaxis(::AbstractArray) = Base.OneTo(1)`, or should it inherit from the parent the way reductions like `sum` do:
```julia
julia> using OffsetArrays

julia> oa = OffsetArray([1 2 3; 4 5 6], 7, 8)
2×3 OffsetArray(::Matrix{Int64}, 8:9, 9:11) with eltype Int64 with indices 8:9×9:11:
 1  2  3
 4  5  6

julia> sum(oa, dims=1)  # trivial axis is 8:8
1×3 OffsetArray(::Matrix{Int64}, 8:8, 9:11) with eltype Int64 with indices 8:8×9:11:
 5  7  9

julia> eachslice(oa, dims=2, drop=false)  # trivial axis is 1:1
1×3 eachslice(OffsetArray(::Matrix{Int64}, 8:9, 9:11), dims = 2, drop = false) of 2-element slices with eltype Int64 with indices Base.OneTo(1)×OffsetArrays.IdOffsetRange(values=9:11, indices=9:11):
 [1, 4]  [2, 5]  [3, 6]
```

simonbyrne commented:
> Should we add `@boundscheck checkbounds(...)` to `getindex(::Slices, ...)`?

Ah, good catch! Can you open an issue (or a PR)?

simonbyrne commented:
> Relatedly, should the new axis be `unitaxis(::AbstractArray) = Base.OneTo(1)`, or should it inherit from the parent the way reductions like `sum` do:

It should probably inherit from the parent?

oschulz commented:
>> Relatedly, should the new axis be unitaxis(::AbstractArray) = Base.OneTo(1)
> It should probably inherit from the parent?

I was wondering about that too - it would be more natural to inherit indexing from the parent, I think.

oschulz commented:
Another thing, while we're taking fixes and changes: I'm [currently building](https://github.com/oschulz/ArrayInterface.jl/blob/sliced-arrays/lib/ArrayInterfaceCore/src/sliced_arrays.jl) the generic sliced array interface we've been talking about (CC @Tokazama).

The idea is for code to be able to flatten a sliced arrays in either it's "native" or a specific order of dimensions (e.g. inner dims first or outer dims first), so that efficient computations (e.g large linalg ops) can be performed on it, before (possibly) re-slicing the result.

On thing that I'm struggling with is that the current slices implementation is not fully type stable in respect to dimension order: We currently have

```julia
typeof(eachslice(A, dims = (1, 2))) != typeof(eachslice(A, dims = (1, 3)))
```

but

```julia
typeof(eachslice(A, dims = (1, 2))) == typeof(eachslice(A, dims = (2, 1)))
```

So it's impossible to decide whether a `permutedims`/`adjoint`  is necessary based on the type of `A::Slices`, which means unslicing operations that need to unslice to a specified order of dimensions can't be type stable (the only workaround I see would be to make a copy afterwards in all cases, which would be wasteful). This is a problem because code that aims to take advantage of sliced arrays will often need to request a specific order of dimensions (e.g. to use linear algebra on the flat array).


While Julia v0.9 isn't branched yet - could we still change `Slices` and make the slicemap a `Val` or so? I think that would open up a lot more possibilities.

Tokazama commented:
In the ArrayInterface code you're using you can use `StaticInt` to get around this sort of issue.

oschulz commented:
> In the ArrayInterface code you're using you can use StaticInt to get around this sort of issue.

WIthin ArrayInterface, yes, but it won't be in the `Slices` object passed by the user, resp. some other code. I don't think a type-stable flatten-as-specified is possible unless the dimension order in `Slices` can be fully inferred. Dimension order will rarely be dynamically chosen during when `Slices` are created, so I think there's no reason not to just encode it in it's type.

@simonbyrne , this is your baby though, what do you think?

oschulz commented:
> @simonbyrne: Can you open an issue (or a PR)?
 
Sure! @simonbyrne , would you be fine with me making the slicemap part of the `Slices` type in that PR as well (see above)?

oschulz commented:
> making the slicemap part of the Slices type in that PR as well (see above)

#44538 would present another way to make the slicemap inferable (CC @Tokazama ).

@simonbyrne, do you have any preferences here?

Tokazama commented:
I'm assuming it would work to put the slicemap in the type just as well as it works to put the permutation in `PermutedDimsArray`. I do think it's odd that we keep building systems where constant propagation is the only way to get something like `PermutedDimsArray(x, perm)` to be type stable, when we know `perm` is a type parameter (not dynamic). We would think it's weird if the official way to construct an array was `Array(T::DataType, sz::Tuple)`, but we keep making exceptions for indexing into dimensions.

oschulz commented:
> we keep building systems where constant propagation is the only way to get something like PermutedDimsArray(x, perm) to be type stable

Indeed, especially since we want to be able to pass such things around.
